src/main/java/io/jenkins/jenkins/plugins/okhttp/api/JenkinsOkHttpClient.java

```diff
@@ -0,0 +1,91 @@
+package io.jenkins.jenkins.plugins.okhttp.api;
+
+import hudson.ProxyConfiguration;
+import jenkins.model.Jenkins;
+import okhttp3.Authenticator;
+import okhttp3.Credentials;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+import okhttp3.Route;
+
+import javax.net.ssl.SSLSocketFactory;
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.net.Proxy;
+import java.net.ProxySelector;
+import java.net.SocketAddress;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.regex.Pattern;
+
+/**
+ * This class allows to take into consideration if Jenkins is running behind a proxy or not and return a proper {@link OkHttpClient}
+ * client.
+ */
+public class JenkinsOkHttpClient {
+
+    private static final Logger LOGGER = Logger.getLogger(JenkinsOkHttpClient.class.getName());
+
+    private JenkinsOkHttpClient() {
+    }
+
+    /**
+     * Generates a new builder for the given base client.
+     * It applies the current {@link Jenkins#proxy} configuration.
+     *
+     * @return a builder to configure the client
+     */
+    public static OkHttpClient.Builder newClientBuilder(OkHttpClient httpClient) {
+        OkHttpClient.Builder reBuild = httpClient.newBuilder();
+        if (Jenkins.get().proxy != null) {
+            final ProxyConfiguration proxy = Jenkins.get().proxy;
+            Proxy proxyServer = new Proxy(Proxy.Type.HTTP, new InetSocketAddress(proxy.name, proxy.port));
+            if (proxy.getUserName() != null) {
+                Authenticator proxyAuthenticator = new Authenticator() {
+                    @Override public Request authenticate(Route route, Response response) throws IOException {
+                        String credential = Credentials.basic(proxy.getUserName(), proxy.getPassword());
+                        return response.request().newBuilder()
+                                       .header("Proxy-Authorization", credential)
+                                       .build();
+                    }
+                };
+                reBuild.proxyAuthenticator(proxyAuthenticator);
+            }
+
+            // TODO: https://github.com/square/okhttp/issues/3787
+            // not sure if this is correct, needs manual testing
+            if (proxy.name.startsWith("https")) {
+                reBuild.socketFactory(SSLSocketFactory.getDefault());
+            }
+            ProxySelector proxySelector = new ProxySelector() {
+                @Override
+                public List<Proxy> select(URI uri) {
+                    final List<Proxy> proxies = new ArrayList<>(1);
+                    String host = uri.getHost();
+                    for (Pattern p : proxy.getNoProxyHostPatterns()) {
```

Comment:
Jenkins no host are not regular expressions. so this will blow up spectaularly with somethign that is valid for a no proxy host but not valid for a regexp, or give you incorrect results.

> Specify host name patterns that shouldn't go through the proxy, one host per line. "*" is the wild card host name (such as "*.jenkins.io" or "www*.jenkins-ci.org")

in other words `*` is not "the thing before multiple times, but any valid character for a host or dns segment name (`a-zA-Z-_`)  (where `_` is only allowed in the hostname part, not the dns segment path)  (and probably but needs checking would exclude the `.` (dot) character so that `*.example.com` would match `foo.example.com` but not `bar.wibble.example.com`)

This code should probably be made a static class so it can be better tested.